File: AfterToken/.tmp_console.py
import subprocess, json, sys

# 用法: python .tmp_console.py [filter_regex]
pattern = sys.argv[1] if len(sys.argv) > 1 else ""

code = r'''
var logEntriesType = System.AppDomain.CurrentDomain.GetAssemblies()
    .SelectMany(a => { try { return a.GetTypes(); } catch { return new System.Type[0]; } })
    .FirstOrDefault(t => t.FullName == "UnityEditor.LogEntries");
var logEntryType = System.AppDomain.CurrentDomain.GetAssemblies()
    .SelectMany(a => { try { return a.GetTypes(); } catch { return new System.Type[0]; } })
    .FirstOrDefault(t => t.FullName == "UnityEditor.LogEntry");
var flags = System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.NonPublic;
var flagsI = System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.NonPublic;
int count = (int)logEntriesType.GetMethod("GetCount", flags).Invoke(null, null);
var entry = System.Activator.CreateInstance(logEntryType);
var getEntry = logEntriesType.GetMethod("GetEntryInternal", flags);
var msgField = logEntryType.GetField("message", flagsI);
var modeField = logEntryType.GetField("mode", flagsI);
var sb = new System.Text.StringBuilder();
int shown = 0;
for (int i = 0; i < count && shown < 40; i++)
{
    getEntry.Invoke(null, new object[] { i, entry });
    string m = (string)msgField.GetValue(entry);
    int mode = (int)modeField.GetValue(entry);
    var line = "[" + mode + "] " + m.Split('\n')[0];
    sb.AppendLine(line);
    shown++;
}
return "total=" + count + "\n" + sb.ToString();
'''
payload = json.dumps({"action": "execute", "code": code})
r = subprocess.run(["python", ".tmp_unity_mcp.py", "execute_code", payload],
                   capture_output=True, text=True, encoding="utf-8", errors="replace")
out = r.stdout
try:
    data = json.loads(out[out.index("{"):])
    print(data.get("data", {}).get("result", out))
except Exception:
    print(out[-2000:])
if r.stderr.strip():
    print("STDERR:", r.stderr[-300:], file=sys.stderr)

File: AfterToken/.tmp_console.cs
var asm = typeof(UnityEditor.Editor).Assembly;
var logEntries = asm.GetType("UnityEditor.LogEntries") ?? asm.GetType("UnityEditorInternal.LogEntries");
if (logEntries == null) return "no type";
var sb = new System.Text.StringBuilder();
var flags = System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Public;
// 先清再统计不行，直接读行
var startGetting = logEntries.GetMethod("StartGettingEntries", flags);
var getEntry = logEntries.GetMethod("GetEntryInternal", flags);
var endGetting = logEntries.GetMethod("EndGettingEntries", flags);
var entryType = asm.GetType("UnityEditor.LogEntry") ?? asm.GetType("UnityEditorInternal.LogEntry");
if (startGetting == null || getEntry == null || entryType == null) return "methods missing";
int total = (int)startGetting.Invoke(null, null);
var entry = System.Activator.CreateInstance(entryType);
int errors = 0;
var msgField = entryType.GetField("message") ?? entryType.GetField("condition");
var modeField = entryType.GetField("mode");
int shown = 0;
for (int i = total - 1; i >= 0 && shown < 8; i--)
{
    var args = new object[]{ i, entry };
    if (!(bool)getEntry.Invoke(null, args)) continue;
    entry = args[1];
    int mode = modeField != null ? (int)modeField.GetValue(entry) : 0;
    // mode bit: Error=1<<0? 实际 Error=1, Assert=2... 用 ScriptCompileError 位 (1<<8?) — 简化：包含 "error CS" 即显示
    string msg = (string)msgField.GetValue(entry) ?? "";
    if (msg.Contains("error CS") || msg.Contains("Error"))
    {
        var firstLine = msg.Split('\n')[0];
        sb.AppendLine(firstLine.Length > 200 ? firstLine.Substring(0,200) : firstLine);
        shown++;
    }
}
endGetting.Invoke(null, null);
return shown == 0 ? "no compile errors in console" : sb.ToString();
